src/main/java/com/iexec/core/chain/DealWatcherService.java

```diff
@@ -175,7 +190,36 @@ void replayDealEvent() {
     private Disposable subscribeToDealEventInRange(BigInteger from, BigInteger to) {
         log.info("Replay Watcher DealEvent started [from:{}, to:{}]",
                 from, (to == null) ? "latest" : to);
-        return iexecHubService.getDealEventObservable(from, to)
+        EthFilter filter = createDealEventFilter(from, to);
+        return iexecHubService.getDealEventObservable(filter)
+                .map(this::schedulerNoticeToDealEvent)
                 .subscribe(dealEvent -> dealEvent.ifPresent(this::onDealEvent));
     }
+
+    EthFilter createDealEventFilter(BigInteger from, BigInteger to) {
+        DefaultBlockParameter fromBlock = DefaultBlockParameter.valueOf(from);
+        DefaultBlockParameter toBlock = DefaultBlockParameterName.LATEST;
+        if (to != null) {
+            toBlock = DefaultBlockParameter.valueOf(to);
+        }
+        EthFilter filter = new EthFilter(fromBlock, toBlock, chainConfig.getHubAddress());
+        BigInteger poolAddressBigInt = Numeric.toBigInt(chainConfig.getPoolAddress());
+        filter.addSingleTopic(EventEncoder.encode(SCHEDULERNOTICE_EVENT));
+        filter.addSingleTopic(Numeric.toHexStringWithPrefixZeroPadded(poolAddressBigInt, 64));
+        return filter;
+    }
+
+    Optional<DealEvent> schedulerNoticeToDealEvent(IexecHubContract.SchedulerNoticeEventResponse schedulerNotice) {
+        dealEventsCount++;
+        BigInteger noticeBlockNumber = schedulerNotice.log.getBlockNumber();
+        if (latestBlockNumberWithDeal.compareTo(noticeBlockNumber) < 0) {
+            latestBlockNumberWithDeal = noticeBlockNumber;
+        }
+        log.info("Received new deal [dealEventsCount:{}, chainDealId:{}]", dealEventsCount, BytesUtils.bytesToString(schedulerNotice.dealid));
+        if (schedulerNotice.workerpool.equalsIgnoreCase(chainConfig.getPoolAddress())) {
```

Comment:
I usually return the error case instead of the successful case in an `if` block. I find it clearer as this is what I'm expecting to read. Something like:
```java
if (!schedulerNotice.workerpool.equalsIgnoreCase(chainConfig.getPoolAddress())) {
    log.warn(...);
    return Optional.empty();
}
return Optional.of(...);
```

Reply:
I personally think error should be the default, and the good path a chance.
(= I like current implem* :) )

Reply:
Well, maybe both of our opinions are right here. I mean, when there are multiple failure cases and a single success, the success usually is the last line of the method. But when there are multiple success success and a single "default" failure case, then this failure case is the last line.
Here, a single success case and a single failure case, both implementations would work.